### PR TITLE
feature/issue-61: Resolve Dimensions with MetaData

### PR DESCRIPTION
**Pull request by nlenssen2013**
Github Issue: #61 

### Description

SNDR and TROPOMI have variables without dimensions that need to be transfered

### Overview of work done

Resolve issues that xarray_enhancements wasn't completely fixing dimensions

### Overview of verification done

_Summarize the testing and verification you've done. This includes unit tests or testing with specific data_

### Overview of integration done

_Explain how this change was integration tested. Provide screenshots or logs if appropriate. An example of this would be a local Harmony deployment._

## PR checklist:

* [x] Linted
* [x] Updated unit tests
* [x] Updated changelog
* [ ] Integration testing

_See [Pull Request Review Checklist](../CONTRIBUTING.md#reviewing) for pointers on reviewing this pull request_

### CODE REVIEW

File: podaac/subsetter/subset.py
Comment (skorper) [inline podaac/subsetter/subset.py:1126]: Can you remove this print statement and any others?

File: tests/test_subset.py
Comment (skorper) [inline tests/test_subset.py:1690]: It seems like this test contains a lot of duplicate code from the subsetter itself -- can you call a function from the subsetter and make assertions on the result rather than doing it this way?

File: podaac/subsetter/subset.py
Comment (skorper) [inline podaac/subsetter/subset.py:969]: Wouldn't `var_dims` already be an empty list? It seems like this line does nothing. Also I'm a little concerned about this breaking functionality with our existing datasets

File: podaac/subsetter/subset.py
Comment (skorper) [inline podaac/subsetter/subset.py:740]: Why did you pull this out of the `if variables` block? Wouldn't this do nothing because the `if var in variables` would always be False?

**Review by skorper — CHANGES_REQUESTED** [review]

File: podaac/subsetter/subset.py
Comment (nlenssen2013) [reply podaac/subsetter/subset.py:740]: Re-added the if variables statement

File: tests/test_subset.py
Comment (nlenssen2013) [reply tests/test_subset.py:1690]: Trying to minimize calling the entire subset.subset method and just calling the subset_with_bbox method. Will just add assertions to the dimensions input vs output

File: podaac/subsetter/subset.py
Comment (nlenssen2013) [reply podaac/subsetter/subset.py]: Yes, I removed the var_dims reassign. Do you have an idea of what existing datasets might break because all the unittests pass?

File: podaac/subsetter/xarray_enhancements.py
Comment (nlenssen2013) [reply podaac/subsetter/xarray_enhancements.py:244]: 1. It is not, to my knowledge. The logic in line 201 needs to be applied to cases with and without _FillValue. A few SNDR variables were not getting picked up in this logic, thus the excess dimensions
2. If I understand correctly you mean the case when a variable has only one of the dimensions that are being subset? /sat_vel in SNDR has dimensions of (atrack,spatial) and after a bbox subset the atrack dimension is lower as expected

Comment (skorper) [reply podaac/subsetter/xarray_enhancements.py:244]: > It is not, to my knowledge. The logic in line 201 needs to be applied to cases with and without _FillValue. A few SNDR variables were not getting picked up in this logic, thus the excess dimensions

Are you planning on making that change before we merge?

Comment (nlenssen2013) [reply podaac/subsetter/xarray_enhancements.py:244]: > > It is not, to my knowledge. The logic in line 201 needs to be applied to cases with and without _FillValue. A few SNDR variables were not getting picked up in this logic, thus the excess dimensions
> 
> Are you planning on making that change before we merge?

The line that you commented on is the change. Was just explaining what the change was doing and what was lacking before

Comment (skorper) [reply podaac/subsetter/xarray_enhancements.py:244]: Are you referring to the following 3 lines where this same logic is applied in the missing FillValue case? https://github.com/podaac/l2ss-py/blob/94f1630d64aa00e14e23584c3d1434efc5e7f618/podaac/subsetter/xarray_enhancements.py#L225-L228

My concern is that the block above is only supposed to be applied in cases where this is no FillValue. When there is FillValue, data outside the provided bounds are replaced with FillValue. If there is no FillValue, what we use the original data, meaning data outside the bounds will still be in the subsetted result, which results in an incomplete subset. It seems like you're applying this logic to cases where there *is* a FillValue, meaning we will get an incomplete subset even when we don't have to. 

> after a bbox subset the atrack dimension is lower as expected

This doesn't necessarily mean the data is fully subset. Our subset operation has two steps:

1. Cut dimensions as much as possible, meaning dimension size itself will shrink. This is the "indexers" logic you see [here](https://github.com/podaac/l2ss-py/blob/94f1630d64aa00e14e23584c3d1434efc5e7f618/podaac/subsetter/xarray_enhancements.py#L53)
2. Within the remaining data, mask values that are not within the bounds. This means data within the now reduced dimensions will be replaced with "nan" if they are outside of the bounds.
   - If the variable has FillValue, replace nan with FillValue
   - If the variable has no FillValue, replace nan with original values

Sorry for the essay, please let me know if this makes sense

Comment (skorper) [reply podaac/subsetter/xarray_enhancements.py]: @nlenssen2013 Thanks for the explanation. I think my concerns have been alleviated and I will approve once you fix the conflict 🙂

**Comment by skorper** [comment]
@frankinspace SonarCloud is reporting code smell for this PR: https://sonarcloud.io/project/issues?resolved=false&severities=CRITICAL&sinceLeakPeriod=true&types=CODE_SMELL&pullRequest=68&id=podaac_l2ss-py&open=AX-3X9FKvbEZitqyh0cy

It seems the issue is cognitive complexity in the `subset` function. This PR only lightly touches this code so I'm thinking we should open a separate issue for simplifying this function and merge this PR/dismiss this SolarCloud failure. Thoughts?

**Comment by frankinspace** [comment]
I was thinking the same thing @skorper. We need to simplify that function but probably not in this PR

**Review by skorper — APPROVED** [review]

**Comment by skorper** [comment]
https://github.com/podaac/l2ss-py/issues/81

**Pull request merged**
